File: Data Structures & Algorithms/house-robber-ii/submission-7.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums)==1:
            return nums[0]

        def dfs(i,nums,memo):
            if i >= len(nums):
                return 0
            if i in memo:
                return memo[i]
            
            memo[i] = max(nums[i]+dfs(i+2,nums,memo),dfs(i+1,nums,memo))
            return memo[i]
        return max(dfs(0,nums[:-1],{}),dfs(1,nums,{}))

File: Data Structures & Algorithms/top-k-elements-in-list/submission-7.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        std::unordered_map<int,int> mp;
        int mx=0;
        for (auto &num: nums){
            mp[num]++;
            mx = std::max(mx,mp[num]);
        }
        std::vector<std::vector<int>> vec(mx+1);
        for (auto &p: mp){
            vec[p.second].push_back(p.first);
        }
        std::vector<int> ans;
        for (auto it=vec.end(); it!=vec.begin();){
            it--;
            for (auto num:*it){
                if (k!=0) {
                    ans.push_back(num);
                    k--;
                }
            }
        }
        return ans;
    }  
};
